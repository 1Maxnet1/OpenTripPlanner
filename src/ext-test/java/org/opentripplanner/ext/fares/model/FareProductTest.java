package org.opentripplanner.ext.fares.model;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.time.Duration;
import java.time.OffsetDateTime;
import java.time.ZonedDateTime;
import java.util.stream.Stream;
import javax.annotation.Nonnull;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.opentripplanner.model.fare.FareMedium;
import org.opentripplanner.model.fare.FareProduct;
import org.opentripplanner.model.fare.RiderCategory;
import org.opentripplanner.test.support.VariableSource;
import org.opentripplanner.transit.model.basic.Money;
import org.opentripplanner.transit.model.framework.FeedScopedId;

class FareProductTest {

  static ZonedDateTime ZDT = OffsetDateTime.parse("2023-03-27T10:44:54+02:00").toZonedDateTime();

  static RiderCategory CATEGORY = new RiderCategory(
    new FeedScopedId("1", "pensioners"),
    "Pensioners",
    null
  );

  static FareMedium MEDIUM = new FareMedium(new FeedScopedId("1", "app"), "App");

  static Stream<Arguments> testCases = Stream.of(
    Arguments.of(fareProduct(null, null, null), ZDT, "1ef42792-8f41-349f-8bfa-8864930fc50d"),
    Arguments.of(
      fareProduct(null, null, null),
      ZDT.plusHours(1),
      "8d8177ca-bbcd-3c1e-9651-cf6fd749fac3"
    ),
    Arguments.of(
      fareProduct(Duration.ofHours(2), CATEGORY, null),
      ZDT,
      "ec757263-3fee-32dc-856d-5035856b656d"
    ),
    Arguments.of(
      fareProduct(Duration.ofHours(2), CATEGORY, MEDIUM),
      ZDT,
      "86b3d1ca-1145-3c36-aa1f-8f2a685234b5"
    )
  );

  @ParameterizedTest
  @VariableSource("testCases")
  void instanceId(FareProduct fareProduct, ZonedDateTime startTime, String expectedInstanceId) {
    var instanceId = fareProduct.uniqueInstanceId(startTime);

    assertEquals(expectedInstanceId, instanceId);
  }

  @Nonnull
  private static FareProduct fareProduct(Duration duration, RiderCategory cat, FareMedium medium) {
    return new FareProduct(
      new FeedScopedId("fares", "daypass"),
      "day pass",
      Money.euros(10),
      duration,
      cat,
      medium
    );
  }
}
