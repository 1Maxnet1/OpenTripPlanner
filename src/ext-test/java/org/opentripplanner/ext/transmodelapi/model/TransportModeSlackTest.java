package org.opentripplanner.ext.transmodelapi.model;

import java.util.List;
import java.util.Map;
import org.junit.Assert;
import org.junit.Test;
import org.opentripplanner.model.modes.TransitMainMode;

public class TransportModeSlackTest {

    @Test
    public void mapToApiList() {
        // Given
        Map<TransitMainMode, Integer> domain = Map.of(
                TransitMainMode.FUNICULAR, 600,
                TransitMainMode.CABLE_CAR, 600,
                TransitMainMode.RAIL, 1800,
                TransitMainMode.AIRPLANE, 3600
        );

        // When
        List<TransportModeSlack> result = TransportModeSlack.mapToApiList(domain);

        Assert.assertEquals(600, result.get(0).slack);
        Assert.assertTrue(result.get(0).modes.contains(TransitMainMode.CABLE_CAR));
        Assert.assertTrue(result.get(0).modes.contains(TransitMainMode.FUNICULAR));

        Assert.assertEquals(1800, result.get(1).slack);
        Assert.assertTrue(result.get(1).modes.contains(TransitMainMode.RAIL));

        Assert.assertEquals(3600, result.get(2).slack);
        Assert.assertTrue(result.get(2).modes.contains(TransitMainMode.AIRPLANE));
    }

    @Test
    public void mapToDomain() {
        // Given
        List<Object> apiSlackInput = List.of(
                Map.of(
                        "slack", 600,
                        "modes", List.of(TransitMainMode.FUNICULAR, TransitMainMode.CABLE_CAR)
                ),
                Map.of(
                        "slack", 1800,
                        "modes", List.of(TransitMainMode.RAIL)
                ),
                Map.of(
                        "slack", 3600,
                        "modes", List.of(TransitMainMode.AIRPLANE)
                )
        );


        Map<TransitMainMode, Integer> result;

        // When
        result = TransportModeSlack.mapToDomain(apiSlackInput);

        // Then
        Assert.assertNull(result.get(TransitMainMode.BUS)  );
        Assert.assertEquals(Integer.valueOf(600), result.get(TransitMainMode.FUNICULAR));
        Assert.assertEquals(Integer.valueOf(600), result.get(TransitMainMode.CABLE_CAR));
        Assert.assertEquals(Integer.valueOf(1800), result.get(TransitMainMode.RAIL));
        Assert.assertEquals(Integer.valueOf(3600), result.get(TransitMainMode.AIRPLANE));
    }
}