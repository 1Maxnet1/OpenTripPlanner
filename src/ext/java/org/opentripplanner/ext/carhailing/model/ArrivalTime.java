package org.opentripplanner.ext.carhailing.model;

import java.time.Duration;

// A class to model estimated arrival times of service from a car hailing company
public record ArrivalTime(
  CarHailingProvider provider,
  String productId,
  String displayName,
  Duration estimatedDuration,
  boolean wheelchairAccessible
) {}
