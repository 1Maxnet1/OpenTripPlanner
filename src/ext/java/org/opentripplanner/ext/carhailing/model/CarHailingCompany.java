package org.opentripplanner.ext.carhailing.model;

public enum CarHailingCompany {
  LYFT,
  NOAPI,
  UBER,
}
