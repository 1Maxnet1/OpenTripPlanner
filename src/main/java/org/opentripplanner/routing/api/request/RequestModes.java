package org.opentripplanner.routing.api.request;

import java.util.Collection;
import java.util.List;
import java.util.Objects;
import javax.annotation.Nonnull;
import org.opentripplanner.transit.model.network.MainAndSubMode;
import org.opentripplanner.util.lang.ToStringBuilder;

public class RequestModes {

  /**
   * The RequestModes is mutable, so we need to keep a private default set of modes.
   */
  private static final RequestModes defaultRequestModes = new RequestModes(
    StreetMode.WALK,
    StreetMode.WALK,
    StreetMode.WALK,
    StreetMode.WALK,
    MainAndSubMode.all()
  );

  @Nonnull
  public StreetMode accessMode;

  @Nonnull
  public StreetMode egressMode;

  @Nonnull
  public StreetMode directMode;

  @Nonnull
  public StreetMode transferMode;

  @Nonnull
  public List<MainAndSubMode> transitModes;

  public RequestModes(
    StreetMode accessMode,
    StreetMode transferMode,
    StreetMode egressMode,
    StreetMode directMode,
    Collection<MainAndSubMode> transitModes
  ) {
    this.accessMode = (accessMode != null && accessMode.access) ? accessMode : StreetMode.NOT_SET;
    this.egressMode = (egressMode != null && egressMode.egress) ? egressMode : StreetMode.NOT_SET;
    this.directMode = directMode != null ? directMode : StreetMode.NOT_SET;
    this.transferMode =
      (transferMode != null && transferMode.transfer) ? transferMode : StreetMode.NOT_SET;
    this.transitModes = transitModes == null ? MainAndSubMode.all() : List.copyOf(transitModes);
  }

  public static RequestModes defaultRequestModes() {
    // Clone default since this class is mutable
    return new RequestModes(
      defaultRequestModes.accessMode,
      defaultRequestModes.transferMode,
      defaultRequestModes.egressMode,
      defaultRequestModes.directMode,
      defaultRequestModes.transitModes
    );
  }

  public boolean contains(StreetMode streetMode) {
    return (
      streetMode.equals(accessMode) ||
      streetMode.equals(egressMode) ||
      streetMode.equals(directMode)
    );
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;

    RequestModes that = (RequestModes) o;

    if (accessMode != that.accessMode) {
      return false;
    }
    if (egressMode != that.egressMode) {
      return false;
    }
    if (directMode != that.directMode) {
      return false;
    }
    if (transferMode != that.transferMode) {
      return false;
    }
    return transitModes.equals(that.transitModes);
  }

  @Override
  public int hashCode() {
    return Objects.hash(accessMode, egressMode, directMode, transferMode, transitModes);
  }

  @Override
  public String toString() {
    return ToStringBuilder
      .of(RequestModes.class)
      .addEnum("accessMode", accessMode)
      .addEnum("egressMode", egressMode)
      .addEnum("directMode", directMode)
      .addEnum("transferMode", transferMode)
      .addCol("transitModes", transitModes)
      .toString();
  }
}
