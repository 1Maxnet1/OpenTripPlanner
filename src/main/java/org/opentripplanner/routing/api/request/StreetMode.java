package org.opentripplanner.routing.api.request;

import org.opentripplanner.framework.doc.DocumentedEnum;

public enum StreetMode implements DocumentedEnum<StreetMode> {
  /**
   * No street mode is set. This option is used if we do not want street routing at all in this part
   * of the search.
   */
  NOT_SET(true, true, true, false, false, false, false, false, false),
  /**
   * Walk only
   */
  WALK(true, true, true, true, false, false, false, false, false),
  /**
   * Bike only
   */
  BIKE(true, true, true, false, true, false, false, false, false),
  /**
   * Bike to a bike parking area, then walk the rest of the way.
   * <p>
   * Direct mode and access mode only.
   */
  BIKE_TO_PARK(true, false, false, true, true, false, false, true, false),
  /**
   * Walk to a bike rental point, bike to a bike rental drop-off point, and walk the rest of the
   * way. This can include bike rental at fixed locations or free-floating services.
   */
  BIKE_RENTAL(true, true, true, true, true, false, true, false, false),
  /**
   * Walk to a scooter rental point, ride a scooter to a scooter rental drop-off point, and walk the
   * rest of the way. This can include scooter rental at fixed locations or free-floating services.
   */
  SCOOTER_RENTAL(true, true, true, true, true, false, true, false, false),
  /**
   * Car only
   * <p>
   * Direct mode only.
   */
  CAR(true, false, false, false, false, true, false, false, false),
  /**
   * Start in the car, drive to a parking area, and walk the rest of the way.
   * <p>
   * Direct mode and access mode only.
   */
  CAR_TO_PARK(true, false, false, true, false, true, false, true, false),
  /**
   * Walk to a pickup point along the road, drive to a drop-off point along the road, and walk the
   * rest of the way. This can include various taxi-services or kiss & ride.
   */
  CAR_PICKUP(true, false, true, true, false, true, false, false, true),
  /**
   * Walk to a car rental point, drive to a car rental drop-off point and walk the rest of the way.
   * This can include car rental at fixed locations or free-floating services.
   */
  CAR_RENTAL(true, true, true, true, false, true, true, false, false),

  /**
   * Using a car hailing app like Uber or Lyft to get to a train station or all the way to the destination.
   */
  CAR_HAILING(true, false, true, false, false, true, false, false, true),

  /**
   * Encompasses all types of on-demand and flexible transportation.
   */
  FLEXIBLE(true, false, true, true, false, false, false, false, false);

  final boolean access;

  final boolean transfer;

  final boolean egress;

  final boolean includesWalking;

  final boolean includesBiking;

  final boolean includesDriving;

  final boolean includesRenting;

  final boolean includesParking;

  final boolean includesPickup;

  StreetMode(
    boolean access,
    boolean transfer,
    boolean egress,
    boolean includesWalking,
    boolean includesBiking,
    boolean includesDriving,
    boolean includesRenting,
    boolean includesParking,
    boolean includesPickup
  ) {
    this.access = access;
    this.transfer = transfer;
    this.egress = egress;
    this.includesWalking = includesWalking;
    this.includesBiking = includesBiking;
    this.includesDriving = includesDriving;
    this.includesRenting = includesRenting;
    this.includesParking = includesParking;
    this.includesPickup = includesPickup;
  }

  public boolean includesWalking() {
    return includesWalking;
  }

  public boolean includesBiking() {
    return includesBiking;
  }

  public boolean includesDriving() {
    return includesDriving;
  }

  public boolean includesRenting() {
    return includesRenting;
  }

  public boolean includesParking() {
    return includesParking;
  }

  public boolean includesPickup() {
    return includesPickup;
  }

  @Override
  public String typeDescription() {
    return "Routing modes on streets, including walking, biking, driving, and car-sharing.";
  }

  @Override
  public String enumValueDescription() {return switch (this){

    case NOT_SET -> "";
    case WALK -> "Walking some or all of the way of the route.";
    case BIKE -> "Cycling for the entirety of the route or taking a bicycle onto the public transport and cycling from the arrival station to the destination. ";
    case BIKE_TO_PARK -> "Leaving the bicycle at the departure station and walking from the arrival station to the destination. <br/> This mode needs to be combined with at least one transit mode (or `TRANSIT`) otherwise it behaves like an ordinary bicycle journey. <br/> _Prerequisite:_ Bicycle parking stations present in the OSM file and visible to OTP by enabling the property `staticBikeParkAndRide` during graph build.";
    case BIKE_RENTAL -> "Taking a rented, shared-mobility bike for part or the entirety of the route. <br/>  _Prerequisite:_ Vehicle positions need to be added to OTP from dynamic data feeds. <br/>   For dynamic bike positions configure an input feed. See [Configuring real-time updaters](UpdaterConfig.md).";
    //TODO: I assume this also needs dynamic data feeds?
    case SCOOTER_RENTAL -> "Walking to a scooter rental point, riding a scooter to a scooter rental drop-off point, and walking the rest of the way. <br/> This can include scooter rental at fixed locations or free-floating services.";
    case CAR -> "Driving your own car the entirety of the route. <br/> If this is combined with `TRANSIT`, it will return routes with a [Kiss & Ride](https://en.wikipedia.org/wiki/Park_and_ride#Kiss_and_ride_/_kiss_and_fly) component. This means that the car is not parked in a permanent parking  area but rather the passenger is dropped off (for example, at an airport) and the driver continues driving the car away from the drop off location.";
    case CAR_TO_PARK -> "Driving a car to the park-and-ride facilities near a station and taking publictransport. <br/> This mode needs to be combined with at least one transit mode (or `TRANSIT`) otherwise, it behaves like an ordinary car journey. <br/> _Prerequisite:_ Park-and-ride areas near the stations need to be present in the OSM input file.";
    case CAR_PICKUP -> "Walking to a pickup point along the road, driving to a drop-off point along the road, and walking the rest of the way. <br/> This can include various taxi-services or kiss & ride.";
    //TODO: I assume this also needs dynamic data feeds?
    case CAR_RENTAL -> "Walk to a car rental point, drive to a car rental drop-off point and walk the rest of the way. <br/> This can include car rental at fixed locations or free-floating services.";
    case CAR_HAILING -> "Using a car hailing app like Uber or Lyft to get to a train station or all the way to the destination.";
    case FLEXIBLE -> "Encompasses all types of on-demand and flexible transportation.";
  };

  }
}
