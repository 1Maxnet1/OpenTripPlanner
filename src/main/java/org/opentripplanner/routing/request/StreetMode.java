package org.opentripplanner.routing.request;

public enum StreetMode {
  /**
   * Walk only
   */
  WALK(true, true),
  /**
   * Bike only
   *
   * Direct mode only. Bike on transit is not yet supported.
   */
  BIKE(false, false),
  /**
   * Bike to a bike parking area, then walk the rest of the way.
   *
   * Direct mode and access mode only.
   */
  // TODO OTP2 Only implemented for direct search
  BIKE_TO_PARK(true, false),
  /**
   * Walk to a bike rental point, bike to a bike rental drop-off point, and walk the rest of the
   * way. This can include bike rental at fixed locations or free-floating services.
   */
  // TODO OTP2 Only implemented for direct search
  BIKE_RENTAL(true, true),
  /**
   * Car only
   *
   * Direct mode only.
   */
  CAR(false, false),
  /**
   * Start in the car, drive to a parking area, and walk the rest of the way.
   *
   * Direct mode and access mode only.
   */
  // TODO OTP2 Only implemented for direct search
  CAR_TO_PARK(true, false),
  /**
   * Walk to a pickup point along the road, drive to a drop-off point along the road,
   * and walk the rest of the way. This can include various taxi-services or kiss & ride.
   */
  // TODO OTP2 Only implemented for direct search
  TAXI(true, true),
  /**
   * Walk to a car rental point, drive to a car rental drop-off point and walk the rest of the way.
   * This can include car rental at fixed locations or free-floating services.
   */
  // TODO OTP2 Not implemented
  CAR_RENTAL(true, true);

  boolean access;

  boolean egress;

  StreetMode(boolean access, boolean egress) {
    this.access = access;
    this.egress = egress;
  }
}
