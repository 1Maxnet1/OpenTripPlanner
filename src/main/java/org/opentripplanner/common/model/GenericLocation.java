package org.opentripplanner.common.model;

import com.google.common.base.Joiner;
import org.locationtech.jts.geom.Coordinate;
import org.opentripplanner.model.FeedScopedId;

import java.io.Serializable;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Class describing a location provided by clients of routing. Used to describe end points
 * (origin, destination) of a routing request as well as any intermediate points that should
 * be passed through.
 * <p>
 * Handles parsing of geospatial information from strings so that it need not be littered through
 * the routing code.
 *
 * @author avi
 */
public class GenericLocation implements Cloneable, Serializable {

    /**
     * The name of the place, if provided. This is pass-through information and does not affect
     * routing in any way.
     */
    public final String label;

    /**
     * Refers to a specific element in the OTP model. This can currently be a Stop or StopCollection.
     */
    public final FeedScopedId placeId;

    /**
     * Coordinates of the location. These can be used by themselves or as a fallback if placeId is
     * not found.
     */
    public final Double lat;

    public final Double lng;

    public GenericLocation(String label, FeedScopedId placeId, Double lat, Double lng) {
        this.label = label;
        this.placeId = placeId;
        this.lat = lat;
        this.lng = lng;
    }


    /**
     * Observed heading if any.
     *
     * Direction of travel in decimal degrees from -180° to +180° relative to
     * true north.
     *
     * 0      = heading true north.
     * +/-180 = heading south.
     */
    public Double heading;

    // Pattern for matching lat,lng strings, i.e. an optional '-' character followed by 
    // one or more digits, and an optional (decimal point followed by one or more digits).
    private static final String _doublePattern = "-{0,1}\\d+(\\.\\d+){0,1}";

    // We want to ignore any number of non-digit characters at the beginning of the string, except
    // that signs are also non-digits. So ignore any number of non-(digit or sign or decimal point).
    // Regex has been rewritten following https://bugs.openjdk.java.net/browse/JDK-8189343
    // from "[^[\\d&&[-|+|.]]]*(" to "[\\D&&[^-+.]]*("
    private static final Pattern _latLonPattern = Pattern.compile("[\\D&&[^-+.]]*(" + _doublePattern
            + ")(\\s*,\\s*|\\s+)(" + _doublePattern + ")\\D*");
    
    private static final Pattern _headingPattern = Pattern.compile("\\D*heading=("
            + _doublePattern + ")\\D*");

    /**
     * Constructs an empty GenericLocation.
     */
    public GenericLocation() {
        this.label = "";
        this.place = "";
    }

    /**
     * Constructs a GenericLocation with coordinates only.
     */
    public GenericLocation(double lat, double lng) {
        this.label = "";
        this.place = "";
        this.lat = lat;
        this.lng = lng;
    }

    /**
     * Constructs a GenericLocation with coordinates only.
     */
    public GenericLocation(Coordinate coord) {
        this(coord.y, coord.x);
    }

    /**
     * Constructs a GenericLocation with coordinates and heading.
     */
    public GenericLocation(double lat, double lng, double heading) {
        this.label = "";
        this.place = "";
        this.lat = lat;
        this.lng = lng;
        this.heading = heading;
    }

    /**
     * Construct from a name, place pair.
     * Parses latitude, longitude data, heading and numeric edge ID out of the place string.
     * Note that if the place string does not appear to contain a lat/lon pair, heading, or edge ID
     * the GenericLocation will be missing that information but will still retain the place string,
     * which will be interpreted during routing context construction as a vertex label within the
     * graph for the appropriate routerId (by StreetVertexIndexServiceImpl.getVertexForLocation()).
     * TODO: Perhaps the interpretation as a vertex label should be done here for clarity.
     */
    public GenericLocation(String label, String place) {
        this.label = label;
        this.place = place;

        if (place == null) {
            return;
        }

        Matcher matcher = _latLonPattern.matcher(place);
        if (matcher.find()) {
            this.lat = Double.parseDouble(matcher.group(1));
            this.lng = Double.parseDouble(matcher.group(4));
        }

        matcher = _headingPattern.matcher(place);
        if (matcher.find()) {
            this.heading = Double.parseDouble(matcher.group(1));
        }

    }

    public GenericLocation(String label, String vertexId, Double lat, Double lng) {
        this.label = label;
        // TODO OTP2 - this.vertexId = vertexId;
        this.lat = lat;
        this.lng = lng;
        this.place = Joiner.on(",").skipNulls().join(vertexId, lat, lng);
    }

    /**
     * Creates the GenericLocation by parsing a "name::place" string, where "place" is a latitude,longitude string or a vertex ID.
     *
     * @param input
     * @return
     */
    public static GenericLocation fromOldStyleString(String input) {
        String name = "";
        String place = input;
        if (input.contains("::")) {
            String[] parts = input.split("::", 2);
            name = parts[0];
            place = parts[1];
        }
        return new GenericLocation(name, place);
    }

    /**
     * Returns true if this.heading is not null.
     * @return
     */
    public boolean hasHeading() {
        return heading != null;
    }

    /** Returns true if this.name is set. */
    public boolean hasName() {
        return label != null && !label.isEmpty();
    }

    /** Returns true if this.place is set. */
    public boolean hasPlace() {
        return place != null && !place.isEmpty();
    }

    /**
     * Returns true if getCoordinate() will not return null.
     * @return
     */
    public boolean hasCoordinate() {
        return this.lat != null && this.lng != null;
    }

    public NamedPlace getNamedPlace() {
        return new NamedPlace(this.label, this.place);
    }

    /**
     * Returns this as a Coordinate object.
     * @return
     */
    public Coordinate getCoordinate() {
        if (this.lat == null || this.lng == null) {
            return null;
        }
        return new Coordinate(this.lng, this.lat);
    }
}
