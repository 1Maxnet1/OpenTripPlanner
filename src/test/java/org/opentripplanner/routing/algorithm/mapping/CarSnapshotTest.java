package org.opentripplanner.routing.algorithm.mapping;

import au.com.origin.snapshots.junit5.SnapshotExtension;
import java.util.Locale;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.api.parallel.ResourceLock;
import org.junit.jupiter.api.parallel.Resources;
import org.opentripplanner.model.GenericLocation;
import org.opentripplanner.routing.api.request.RequestModes;
import org.opentripplanner.routing.api.request.RoutingRequest;
import org.opentripplanner.routing.api.request.StreetMode;

@ExtendWith(SnapshotExtension.class)
@ResourceLock(Resources.LOCALE)
public class CarSnapshotTest extends SnapshotTestBase {

  private static final Locale DEFAULT_LOCALE = Locale.getDefault();

  static GenericLocation p1 = new GenericLocation(
    "NW Pettygrove Ave. & NW 24th Ave. (P1)",
    null,
    45.53261,
    -122.70075
  );

  static GenericLocation p2 = new GenericLocation(
    "NW Marshall St. & NW 24th Ave. (P2)",
    null,
    45.53046,
    -122.70067
  );

  @BeforeAll
  public static void beforeClass() {
    Locale.setDefault(Locale.US);
    loadGraphBeforeClass(false);
  }

  @AfterAll
  public static void afterClass() {
    Locale.setDefault(DEFAULT_LOCALE);
  }

  @DisplayName("Direct CAR_TO_PARK")
  @Test
  public void directCarPark() {
    RoutingRequest request = createTestRequest(2009, 10, 21, 16, 10, 0);

    request.modes =
      RequestModes.of().withDirectMode(StreetMode.CAR_TO_PARK).clearTransitModes().build();
    request.from = p1;
    request.to = p2;

    expectArriveByToMatchDepartAtAndSnapshot(request);
  }
}
