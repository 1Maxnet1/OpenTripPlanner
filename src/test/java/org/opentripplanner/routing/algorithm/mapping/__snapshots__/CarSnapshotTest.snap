org.opentripplanner.routing.algorithm.mapping.CarSnapshotTest.directCarPark=[
  [
    {
      "arrivedAtDestinationWithRentedBicycle": false,
      "duration": 253,
      "elevationGained": 0.0,
      "elevationLost": 0.0,
      "endTime": "2009-10-21T23:14:13.000+00:00",
      "fare": {
        "details": { },
        "fare": { }
      },
      "generalizedCost": 467,
      "legs": [
        {
          "agencyTimeZoneOffset": -25200000,
          "arrivalDelay": 0,
          "departureDelay": 0,
          "distance": 181.36,
          "endTime": "2009-10-21T23:11:44.000+00:00",
          "from": {
            "departure": "2009-10-21T23:10:00.000+00:00",
            "lat": 45.53261,
            "lon": -122.70075,
            "name": "NW Pettygrove Ave. & NW 24th Ave. (P1)",
            "vertexType": "NORMAL"
          },
          "generalizedCost": 181,
          "interlineWithPreviousLeg": false,
          "legGeometry": {
            "length": 5,
            "points": "ya|tGv~{kV??nCEB|Dn@@"
          },
          "mode": "CAR",
          "pathway": false,
          "realTime": false,
          "rentedBike": false,
          "route": "",
          "startTime": "2009-10-21T23:10:00.000+00:00",
          "steps": [
            {
              "absoluteDirection": "WEST",
              "area": false,
              "bogusName": false,
              "distance": 0.72,
              "elevation": "",
              "lat": 45.5326129,
              "lon": -122.7007501,
              "relativeDirection": "DEPART",
              "stayOn": false,
              "streetName": "Northwest Pettygrove Street",
              "walkingBike": false
            },
            {
              "absoluteDirection": "SOUTH",
              "area": false,
              "bogusName": false,
              "distance": 80.22,
              "elevation": "",
              "lat": 45.5326127,
              "lon": -122.7007593,
              "relativeDirection": "LEFT",
              "stayOn": false,
              "streetName": "Northwest 24th Avenue",
              "walkingBike": false
            },
            {
              "absoluteDirection": "WEST",
              "area": false,
              "bogusName": false,
              "distance": 74.04,
              "elevation": "",
              "lat": 45.5318916,
              "lon": -122.7007283,
              "relativeDirection": "RIGHT",
              "stayOn": false,
              "streetName": "Northwest Overton Street",
              "walkingBike": false
            },
            {
              "absoluteDirection": "SOUTH",
              "area": false,
              "bogusName": true,
              "distance": 26.38,
              "elevation": "",
              "lat": 45.5318763,
              "lon": -122.7016786,
              "relativeDirection": "LEFT",
              "stayOn": false,
              "streetName": "service road",
              "walkingBike": false
            }
          ],
          "to": {
            "arrival": "2009-10-21T23:11:44.000+00:00",
            "departure": "2009-10-21T23:10:44.000+00:00",
            "lat": 45.5314019,
            "lon": -122.7016931,
            "name": "P+R (#1)",
            "vehicleParking": {
              "entranceId": "OSM:OSMWay/-102488/osm:node:-102437",
              "entranceName": "P+R (#1)",
              "hasAnyCarPlaces": true,
              "hasBicyclePlaces": false,
              "hasCarPlaces": true,
              "hasWheelchairAccessibleCarPlaces": false,
              "id": "OSM:OSMWay/-102488",
              "name": "P+R",
              "realtime": false,
              "tags": [
                "osm:amenity=parking"
              ]
            },
            "vertexType": "BIKEPARK"
          },
          "transitLeg": false,
          "walkingBike": false
        },
        {
          "agencyTimeZoneOffset": -25200000,
          "arrivalDelay": 0,
          "departureDelay": 0,
          "distance": 184.06,
          "endTime": "2009-10-21T23:14:13.000+00:00",
          "from": {
            "arrival": "2009-10-21T23:11:44.000+00:00",
            "departure": "2009-10-21T23:10:44.000+00:00",
            "lat": 45.5314019,
            "lon": -122.7016931,
            "name": "P+R (#1)",
            "vehicleParking": {
              "entranceId": "OSM:OSMWay/-102488/osm:node:-102437",
              "entranceName": "P+R (#1)",
              "hasAnyCarPlaces": true,
              "hasBicyclePlaces": false,
              "hasCarPlaces": true,
              "hasWheelchairAccessibleCarPlaces": false,
              "id": "OSM:OSMWay/-102488",
              "name": "P+R",
              "realtime": false,
              "tags": [
                "osm:amenity=parking"
              ]
            },
            "vertexType": "BIKEPARK"
          },
          "generalizedCost": 285,
          "interlineWithPreviousLeg": false,
          "legGeometry": {
            "length": 5,
            "points": "gz{tGrd|kVn@@CgElCEB?"
          },
          "mode": "WALK",
          "pathway": false,
          "realTime": false,
          "rentedBike": false,
          "route": "",
          "startTime": "2009-10-21T23:10:44.000+00:00",
          "steps": [
            {
              "absoluteDirection": "SOUTH",
              "area": false,
              "bogusName": true,
              "distance": 26.38,
              "elevation": "",
              "lat": 45.5314019,
              "lon": -122.7016931,
              "relativeDirection": "CONTINUE",
              "stayOn": false,
              "streetName": "path",
              "walkingBike": false
            },
            {
              "absoluteDirection": "EAST",
              "area": false,
              "bogusName": false,
              "distance": 77.46,
              "elevation": "",
              "lat": 45.5311647,
              "lon": -122.7017004,
              "relativeDirection": "LEFT",
              "stayOn": false,
              "streetName": "Northwest Northrup Street",
              "walkingBike": false
            },
            {
              "absoluteDirection": "SOUTH",
              "area": false,
              "bogusName": false,
              "distance": 80.22,
              "elevation": "",
              "lat": 45.531181,
              "lon": -122.7007063,
              "relativeDirection": "RIGHT",
              "stayOn": false,
              "streetName": "Northwest 24th Avenue",
              "walkingBike": false
            }
          ],
          "to": {
            "arrival": "2009-10-21T23:14:13.000+00:00",
            "lat": 45.53046,
            "lon": -122.70067,
            "name": "NW Marshall St. & NW 24th Ave. (P2)",
            "vertexType": "NORMAL"
          },
          "transitLeg": false,
          "walkingBike": false
        }
      ],
      "startTime": "2009-10-21T23:10:00.000+00:00",
      "tooSloped": false,
      "transfers": 0,
      "transitTime": 0,
      "waitingTime": -60,
      "walkDistance": 365.42,
      "walkLimitExceeded": false,
      "walkTime": 313
    }
  ]
]